File: USACO/Contests/1819_1P/Itout.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define pii pair<int, int>
#define pil pair<int, ll>

const int mxN=1e5+1;
int n, a[mxN];
ll k;
pil dp[mxN], ft[mxN+1];
vector<pii> b[mxN+1];
bool inlis[mxN];

pil operator+(const pil &a, const pil &b) {
	if(a.first>b.first)
		return a;
	if(a.first<b.first)
		return b;
	return {a.first, min(a.second+b.second, k)};
}

//fenwick tree update
void upd(int i, pil x) {
	for(i=n-i; i<=n; i+=i&-i)
		ft[i]=ft[i]+x;
}

//fenwick tree query
pil qry(int i) {
	pil r={};
	for(i=n-i; i; i-=i&-i)
		r=r+ft[i];
	return r;
}

int main() {
	ifstream cin("itout.in");
	ofstream cout("itout.out");
	
	//input
	//add a 0 in front to make things easier
	cin >> n >> k, ++n;
	for(int i=1; i<n; ++i)
		cin >> a[i];
	//do the normal LIS dp
	for(int i=n-1; i>=0; --i) {
		dp[i]=qry(a[i]);
		dp[i]=pil(dp[i].first+1, dp[i].second)+pil(1, 1);
		upd(a[i], dp[i]);
		//add element to bucket according to length of LIS
		b[dp[i].first].push_back({a[i], i});
	}
	//restore kth largest LIS
	for(int i=1; i<=n; ++i)
		sort(b[i].begin(), b[i].end(), greater<pii>());
	for(int i=0; ; ) {
		inlis[a[i]]=1;
		if(dp[i].first==1)
			break;
		//iterate over possible next elements
		for(pii p : b[dp[i].first-1]) {
			int j=p.second;
			if(j<i)
				continue;
			if(dp[j].second>=k) {
				i=j;
				break;
			} else
				k-=dp[j].second;
		}
	}
	//output the complement
	cout << n-dp[0].first << "\n";
	for(int i=0; i<n; ++i)
		if(!inlis[i])
			cout << i << "\n";
}

File: USACO/Contests/1819_1P/Itout.java
import java.io.*;
import java.util.*;

public class Itout {
	static final boolean stdin = false;
	static final String filename = "itout";
	static Reader in;
	static PrintWriter out;
	
	static class Solver {
		Solver() {
			//input
			//add a 0 in front to make things easier
			int n=in.nextInt()+1;
			long k=in.nextLong();
			final int[] a = new int[n+1];
			for(int i=1; i<n; ++i)
				a[i]=in.nextInt();
			//do the normal LIS dp
			PIL[] dp = new PIL[n];
			FenwickTree ft = new FenwickTree(n);
			List<Integer>[] b = new List[n+1];
			for(int i=1; i<=n; ++i)
				b[i] = new ArrayList<Integer>();
			for(int i=n-1; i>=0; --i) {
				dp[i]=ft.qry(a[i]);
				dp[i]=add(new PIL(dp[i].a+1, dp[i].b), new PIL(1, 1));
				ft.upd(a[i], dp[i]);
				//add element to bucket according to length of LIS
				b[dp[i].a].add(i);
			}
			//restore kth largest LIS
			for(int i=1; i<=n; ++i) {
				Collections.sort(b[i], new Comparator<Integer>() {
					public int compare(Integer i, Integer j) {
						return -(a[i]-a[j]);
					}
				});
			}
			boolean[] inlis = new boolean[n];
			for(int i=0; ; ) {
				inlis[a[i]]=true;
				if(dp[i].a==1)
					break;
				//iterate over possible next elements
				for(Integer j : b[dp[i].a-1]) {
					if(j<i)
						continue;
					if(dp[j].b>=k) {
						i=j;
						break;
					} else
						k-=dp[j].b;
				}
			}
			//output the complement
			out.println(n-dp[0].a);
			for(int i=0; i<n; ++i)
				if(!inlis[i])
					out.println(i);
		}
		
		static PIL add(PIL a, PIL b) {
			if(a.a>b.a)
				return new PIL(a.a, a.b);
			if(a.a<b.a)
				return new PIL(b.a, b.b);
			return new PIL(a.a, Math.min(a.b+b.b, (long)1e18));
		}
		
		static class FenwickTree {
			PIL[] a;
			FenwickTree(int n) {
				a = new PIL[n+1];
				for(int i=1; i<=n; ++i)
					a[i] = new PIL(0, 0);
			}
			void upd(int i, PIL x) {
				for(i=a.length-1-i; i<a.length; i+=i&-i)
					a[i]=add(a[i], x);
			}
			PIL qry(int i) {
				PIL r = new PIL(0, 0);
				for(i=a.length-1-i; i>0; i-=i&-i)
					r=add(r, a[i]);
				return r;
			}
		}
		
		static class PIL {
			int a;
			long b;
			PIL(int a, long b) {
				this.a=a;
				this.b=b;
			}
		}
	}
	
	public static void main(String[] args) throws Exception {
		if(stdin) {
			in = new Reader();
			out = new PrintWriter(System.out);
		} else {
			in = new Reader(filename+".in");
			out = new PrintWriter(filename+".out");
		}
		new Solver();
		out.close();
	}
	
	static class Reader {
		BufferedReader br;
		StringTokenizer st;
		Reader() {
			br = new BufferedReader(new InputStreamReader(System.in));
		}
		Reader(String filename) throws Exception {
			br = new BufferedReader(new FileReader(filename));
		}
		String next() {
			while(st==null||!st.hasMoreTokens()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch(Exception e) {}
			}
			return st.nextToken();
		}
		int nextInt() {
			return Integer.parseInt(next());
		}
		long nextLong() {
			return Long.parseLong(next());
		}
	}
}